### ISSUE TITLE
Object values with references don't work with outputReferences

### ISSUE THREAD

**Issue by dbanksdesign** [amzn/style-dictionary#615]
Hey, @dbanksdesign. Thank you so much for the examples. 

I don't understand enough about the build lifecycle to recommend a solve, but have you tested `transitive: true` with `outputReferences: true`? In this case the `value` is an `Object` and not a `String` and an error is thrown.

* What mechanism would be required if you wanted to pass a reference through a transform?
* Does it even make sense to generate references for custom values?

https://github.com/amzn/style-dictionary/blob/7a28f40b7f44b37e565b1360683b60268d044e9e/examples/advanced/transitive-transforms/sd.config.js#L53

```js
options: {
  outputReferences: true,
},
```

**Output**

```sh
⋊> ~/g/s/e/a/transitive-transforms on 3.0 ⨯ yarn build                                                                                                                                                                              14:04:33
yarn run v1.22.10
$ style-dictionary build --config ./sd.config.js

css
{ h: '{hue.red}', s: '{saturation.1}', l: '{lightness.1}' }
/Users/c0h04ba/git/style-dictionary/examples/advanced/transitive-transforms/node_modules/style-dictionary/lib/utils/references/getReferences.js:45
  value.replace(regex, function(match, variable) {
        ^

TypeError: value.replace is not a function
    at Object.getReferences (/Users/c0h04ba/git/style-dictionary/examples/advanced/transitive-transforms/node_modules/style-dictionary/lib/utils/references/getReferences.js:45:9)
    at sorter (/Users/c0h04ba/git/style-dictionary/examples/advanced/transitive-transforms/node_modules/style-dictionary/lib/common/formatHelpers/sortByReference.js:38:34)
    at Array.sort (<anonymous>)
    at formattedVariables (/Users/c0h04ba/git/style-dictionary/examples/advanced/transitive-transforms/node_modules/style-dictionary/lib/common/formatHelpers/formattedVariables.js:55:40)
    at Object.css/variables [as format] (/Users/c0h04ba/git/style-dictionary/examples/advanced/transitive-transforms/node_modules/style-dictionary/lib/common/formats.js:47:7)
    at buildFile (/Users/c0h04ba/git/style-dictionary/examples/advanced/transitive-transforms/node_modules/style-dictionary/lib/buildFile.js:101:37)
    at /Users/c0h04ba/git/style-dictionary/examples/advanced/transitive-transforms/node_modules/style-dictionary/lib/buildFiles.js:33:7
    at Array.forEach (<anonymous>)
    at buildFiles (/Users/c0h04ba/git/style-dictionary/examples/advanced/transitive-transforms/node_modules/style-dictionary/lib/buildFiles.js:31:18)
    at Object.buildPlatform (/Users/c0h04ba/git/style-dictionary/examples/advanced/transitive-transforms/node_modules/style-dictionary/lib/buildPlatform.js:62:3)
error Command failed with exit code 1.
```

_Originally posted by @chemoish in https://github.com/amzn/style-dictionary/issues/461#issuecomment-828784418_

**Comment by dbanksdesign** [comment]
The issue isn't really with transitive transforms per se, but rather that `outputReferences` doesn't work when a token's value is an object that has references in it. Style Dictionary can resolve these just fine normally, but the `getReferences` function only works on a string.

**Comment by dbanksdesign** [comment]
Also format helpers like `createPropertyFormatter` expect `original.value` to be a string

**Comment by dbanksdesign** [comment]
**note to self: edit this, right now it is stream of conscious**

The issue isn't about `outputReferences` either... If you have a token which has an object for a value (without references) and try to use it in a format like `css/variables` it won't fail, but it would output `--color-red: [object Object];` which is not valid. `outputReferences` makes this issue more explicit by actually failing, which is probably a good thing actually. 

The underlying issue is `createPropertyFormatter` doesn't work with token values that are objects. That is unless you already transform the value to be a string. But in that case the `outputReferences` wouldn't work still because it tries to run a `.replace` on the original value, which is an object in this case. 

Another issue is that `dictionary.usesReference` is run on the *original* value because it will contain the reference. But if you have a value that is an object and transform it into a string, then the original value will be an object and the value will be a string and you won't know how to replace certain parts of the string with the reference. 

One "fix" is to say formats that use `createPropertyFormatter` (built-in or custom) can't handle values that are objects. Because `outputReferences` happens in a format, having a value object is incompatible with `outputReferences`. We can stop the error from happening by doing a simple check if the value is a string before trying to do `.replace` on it. Then we would get rid of the error, but ...

**Comment by dbanksdesign** [comment]
Finally got my head around this to create a fix. With some caveats:

* Token value objects **must** be transformed into a string before formatting. This should already be the case or you will have `[object Object]` in your output files. Using `outputReferences` just made this error more apparent. 
* Token value objects **must** reference another token, because a non-token reference won't have a `name`. This will not *break* the build, but `outputReferences` will skip over trying to use the reference. 
* Transformed value of a token value object must contain the reference *values*. If you transform the token's value so that the referenced value is no longer in that string (like transforming HSL -> hex), it will be skipped by `outputReferences`. Similarly it will build just fine, but won't output the reference. 

Now there shouldn't be any **errors**, but depending on the points above a token might not be able to use `outputReferences`

**Comment by chazzmoney** [comment]
Fixed in #623
